Tests/MongoSwiftTests/DocumentTests.swift

```diff
@@ -583,4 +585,116 @@ final class DocumentTests: MongoSwiftTestCase {
+
+            // doc.first
+            let firstElem = doc[doc.startIndex]
+            expect(doc.first?.key).to(equal(firstElem.key))
+            expect(doc.first?.value).to(bsonEqual(firstElem.value))
+
+            // doc.distance
+            expect(doc.distance(from: doc.startIndex, to: doc.endIndex)).to(equal(doc.count))
+            expect(doc.distance(from: doc.index(after: doc.startIndex), to: doc.endIndex)).to(equal(doc.count - 1))
+
+            // doc.formIndex
+            var firstIndex = 0
+            doc.formIndex(after: &firstIndex)
+            expect(firstIndex).to(equal(doc.index(after: doc.startIndex)))
+
+            // doc.index(offsetBy:), doc.index(offsetBy:,limitedBy:)
+            expect(doc.index(doc.startIndex, offsetBy: 2)).to(equal(doc.endIndex))
+            expect(doc.index(doc.startIndex, offsetBy: 2, limitedBy: doc.endIndex)).to(equal(doc.endIndex))
+            expect(doc.index(doc.startIndex, offsetBy: 99, limitedBy: 1)).to(beNil())
+
+            // firstIndex(where:); This line is commented out because Travis currently builds on 9.4, but this needs 10+
+//            expect(doc.firstIndex { $0.key == "a" && bsonEquals($0.value, 3) }).to(equal(doc.startIndex))
+        }
+
+        fileprivate static func testMutators() throws {
+            var doc: Document = ["a": 3, "b": 2, "c": 5, "d": 4]
+
+            // doc.removeFirst
+            let firstElem = doc.removeFirst()
+            expect(firstElem.key).to(equal("a"))
+            expect(firstElem.value).to(bsonEqual(3))
+            expect(doc).to(equal(["b": 2, "c": 5, "d": 4]))
+            expect(doc).to(haveCorrectCount())
+
+            // doc.removeFirst(k:)
+            doc.removeFirst(2)
+            expect(doc).to(equal(["d": 4]))
+            expect(doc).to(haveCorrectCount())
+
+            // doc.popFirst
+            let lastElem = doc.popFirst()
+            expect(lastElem?.key).to(equal("d"))
+            expect(lastElem?.value).to(bsonEqual(4))
+            expect(doc).to(equal([:]))
+            expect(doc).to(haveCorrectCount())
+
+            // doc.merge
+            let newDoc: Document = ["e": 4, "f": 2]
+            try doc.merge(newDoc)
+            expect(doc).to(haveCorrectCount())
+        }
+
+        fileprivate static func testPrefixSuffix() {
+            let doc: Document = ["a": 3, "b": 2, "c": 5, "d": 4, "e": 3]
+
+            let upToPrefixDoc = doc.prefix(upTo: 2)
+            let throughPrefixDoc = doc.prefix(through: 1)
+            let suffixDoc = doc.suffix(from: 1)
+
+            // doc.prefix(upTo:)
+            expect(upToPrefixDoc).to(equal(["a": 3, "b": 2]))
+            expect(doc).to(haveCorrectCount())
+
+            // doc.prefix(through:)
+            expect(throughPrefixDoc).to(equal(["a": 3, "b": 2]))
+            expect(doc).to(haveCorrectCount())
+
+            // doc.suffix
+            expect(suffixDoc).to(equal(["b": 2, "c": 5, "d": 4, "e": 3]))
+            expect(doc).to(haveCorrectCount())
```

Comment:
It looks like `haveCorrectCount()` simply re-computes a count with `bson_count_keys()`. Would it make more sense to assert our own literal here? In this case, I would expect `doc.count` to equal `5` since there are five unique keys in the document.

Side note: I'm not sure why the `haveCorrectCount()` checks are repeated in this test, as the three sequences are created immediately after `doc` is assigned. The only thing between these `haveCorrectCount()` assertions are equality checks on those sequences, which I wouldn't expect to change the state at all.

Reply:
That approach would work too, I just chose this approach because it lets the programmer not have to worry about actually keeping track of the document state throughout the test case. It's just "easier" to let it do some extra work and compute it for you each time.

The reason these checks exist is because we are maintaining a stateful `count` property on the Document. This `count` property could actually be incorrect, but the `Document` would still pass the equality check because we don't check to see that two documents have the same value for `count` when comparing them for equality.

In other words, since a simple `Document` to `Document` check does not check for `count` equality, and we can't necessarily assume that our logic for incrementing/decrementing the count is correct, we need to explicitly check that the `count` is indeed the correct amount by itself.

Reply:
Would the internal counts actually change between these assertions, though? If I follow your logic, it's the initial calls to `prefix()` and `suffix()` that might alter the count. I don't understand why the count would possibly change between assertions of subsequences that you've already computed at the top of this test method, though.

Should these assertions actually be calling `haveCorrectCount()` on the subsequences themselves instead of repeating the assertion on `doc`?

Reply:
Ah! I see! I actually did not notice this, I see the issue. It _should_ be calling this on `suffixDoc` and friends, not `doc`. This was a mistake. Thanks!